### PR TITLE
ML-23 Feat/mapping spike

### PR DESCRIPTION
**Pull request by dhansak79**
# Site Details Map Component

## Description

This PR introduces a comprehensive interactive map component for displaying site details in the marine licensing application. The implementation provides users with a visual representation of submitted coordinates and site boundaries, significantly enhancing the user experience during the site details review process.

### Key Features

- **Interactive OpenLayers Map**: Displays submitted coordinates with zoom and pan functionality
- **Multiple Coordinate System Support**: Handles both WGS84 and OSGB36 coordinate systems
- **Dynamic Shape Rendering**: Shows circular site boundaries and point coordinates
- **File Upload Visualization**: Renders coordinates from uploaded KML/GeoJSON files
- **Responsive Design**: Integrates seamlessly with GOV.UK Design System

### Technical Implementation

- **Modular Architecture**: 8 single-responsibility classes following Clean Code principles
- **Dynamic Module Loading**: Efficient OpenLayers module imports to reduce bundle size
- **Comprehensive Test Coverage**: new tests across all components, verified with Stryker mutation testing

## Checklist

Before submitting the PR, ensure the following:

- [x] I have run all tests locally and confirmed they pass.
- [x] I have added tests that cover new functionality or changes (if applicable).
- [ ] I have updated the documentation, including README.md and/or other relevant files.

## Related Tickets/Issues

- Implements interactive map visualisation for marine licensing site details
- Addresses user feedback requesting visual coordinate validation
- Supports both manual coordinate entry and file upload workflows

## Changes

**Infrastructure:**

- Webpack configuration for OpenLayers CSS bundling
- SCSS styling integrated with GOV.UK Design System
- Dynamic CSS loading for performance optimization

**Testing:**

- **Unit tests** covering all map functionality
- **Mutation testing** implemented with Stryker for critical algorithms
- **Edge case coverage** for coordinate validation and conversion
- **Integration tests** for map component coordination

### Key Areas to Focus On

1. **Map Integration**: Review how the map integrates with existing review templates
2. **Coordinate Handling**: Verify accuracy of coordinate parsing and conversion
3. **Performance**: Check that CSS/JS only loads on pages with maps

### Testing Instructions

1. **Manual Coordinate Entry**:
   - Navigate to site details → coordinate entry
   - Submit WGS84 coordinates (e.g., 51.489676, -0.231530)
   - Submit OSGB36 coordinates (e.g. 123456, 654321)
   - Verify map displays on review page with circular boundary

2. **File Upload**:
   - Upload a KML/GeoJSON file with coordinates
   - Verify map displays uploaded coordinate points

### Dependencies

- **OpenLayers**: Main mapping library
- **Proj4js**: Coordinate system transformations

No breaking changes to existing functionality.

### Shapefile Upload
image

### KML Upload
image

### WGS84 Circle
image

### OSGB36 Circle
image

## Mutation Testing Results

`file-upload/controller.js`

image

`site-details-map`

image

The OpenLayersModuleLoader contains minimal business logic (just dynamic imports and return statements) and is more effectively tested through integration tests that verify the dependency injection contract works correctly - covered by journey tests that confirm the map is rendered.

## Journey Test Results **NOTE: new tests for the maps will be added in a future PR**

image

The 2 failing tests are known and are being fixed on a separate PR by @jonnywyatt

### CODE REVIEW

File: src/client/javascripts/site-details-map/index.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      this.showError()
+      return 'error'
+    }
+  }
+
+  /**
+   * Display manual coordinates (point or circle)
+   * @param {object} siteDetails - Site details with manual coordinates
+   */
+  displayManualCoordinates(siteDetails) {
```
Comment (jonnywyatt) [inline src/client/javascripts/site-details-map/index.js:118]: This could move into `SiteVisualiser`?

**Review by jonnywyatt — COMMENTED** [review]
Great job to structure the code so clearly 👏 
I've just commented on the flow and structure as I don't understand how the projections / co-ordinate conversion works yet.

File: src/client/stylesheets/components/site-details-map/_site-details-map.scss [hunk 2]
```diff
@@ -0,0 +1,11 @@
+.app-site-details-map {
+  width: 100%;
+  aspect-ratio: 1 / 1;
+  border: 1px solid #b1b4b6;
+}
+
+// Top align map view key in summary list
+.govuk-summary-list__key--top-aligned {
+  // stylelint-disable-next-line declaration-no-important
+  vertical-align: top !important;
```
Comment (jonnywyatt) [inline src/client/stylesheets/components/site-details-map/_site-details-map.scss:10]: This should work without !important

File: src/server/exemption/site-details/review-site-details/file-upload-review.njk [hunk 3]
```diff
@@ -68,6 +76,11 @@
         }) }}
       {% endif %}
 
+      <!-- Site Details Data for JavaScript -->
+      <script type="application/json" id="site-details-data">
+        {{ siteDetailsData | safe }}
```
Comment (jonnywyatt) [inline src/server/exemption/site-details/review-site-details/file-upload-review.njk:81]: Could [dump](https://mozilla.github.io/nunjucks/templating.html#dump) work instead?

File: src/server/exemption/site-details/review-site-details/index.njk [hunk 4]
```diff
@@ -5,6 +5,14 @@
 {% from "govuk/components/button/macro.njk" import govukButton %}
 {% from "cancel-button/macro.njk" import appCancelButton %}
 
+{% block pageTitle %}
+  Review site details | {{ serviceName }}
+{% endblock %}
+
+{% block extraCSS %}
```
[the remark this answers is not in the document]
Comment (dhansak79) [reply src/server/exemption/site-details/review-site-details/index.njk]: Done

File: package.json [hunk 5]
```diff
@@ -2,7 +2,11 @@
   "name": "marine-licensing-frontend",
   "version": "0.0.0",
   "description": "CDP Frontend Template",
-  "sideEffects": false,
+  "sideEffects": [
```
[the remark this answers is not in the document]
Comment (dhansak79) [reply package.json]: fixed

File: src/client/javascripts/site-details-map/index.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      null
+    )
+
+    this.map = this.mapFactory.createMap(this.$root, this.options, vectorLayer)
+
+    this.siteVisualizer.map = this.map
+    this.loadAndDisplaySiteDetails()
+  }
+
+  destroy() {
```
[the remark this answers is not in the document]
Comment (dhansak79) [reply src/client/javascripts/site-details-map/index.js]: Good spot, was only being called from tests - thanks

File: src/client/javascripts/site-details-map/index.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    const olModules = await this.moduleLoader.loadModules()
+
+    if (this.destroyed) {
+      return
+    }
+
+    this.mapFactory = new MapFactory(olModules)
+    const { vectorSource, vectorLayer } = this.mapFactory.createMapLayers()
+    const geoJSONFormat = this.mapFactory.initializeGeoJSONFormat(
+      olModules.GeoJSON
```
[the remark this answers is not in the document]
Comment (dhansak79) [reply src/client/javascripts/site-details-map/index.js]: fixed

File: src/client/javascripts/site-details-map/index.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    this.siteVisualizer = new SiteVisualizer(
+      olModules,
+      vectorSource,
+      geoJSONFormat,
+      null
+    )
+
+    this.map = this.mapFactory.createMap(this.$root, this.options, vectorLayer)
+
+    this.siteVisualizer.map = this.map
```
[the remark this answers is not in the document]
Comment (dhansak79) [reply src/client/javascripts/site-details-map/index.js]: All dependencies are now provided at construction time

File: src/client/javascripts/application.js [hunk 9]
```diff
@@ -28,6 +29,14 @@ document.addEventListener('DOMContentLoaded', () => {
     // eslint-disable-next-line no-new
     new AddAnotherPoint(element)
   })
+
+  const mapElements = document.querySelectorAll(
+    '[data-module="site-details-map"]'
+  )
+  mapElements.forEach((element) => {
+    // eslint-disable-next-line no-new
```
[the remark this answers is not in the document]
Comment (dhansak79) [reply src/client/javascripts/application.js:37]: I'll leave this for now but could be discussed with the team?

File: src/client/javascripts/site-details-map/index.js (hunk 1 above)
Comment (dhansak79) [reply src/client/javascripts/site-details-map/index.js]: Yep, done.

File: src/server/exemption/site-details/review-site-details/file-upload-review.njk (hunk 3 above)
Comment (dhansak79) [reply src/server/exemption/site-details/review-site-details/file-upload-review.njk:82]: The data is already a json string so using dump would wrap it in quotes.

We could pass the raw object from the controller.

I will leave this for now but you can change when you pick up the ticket is you want to.

File: src/client/javascripts/application.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (vasgan25) [reply src/client/javascripts/application.js:37]: we will look to introduce it in the future

**Review by jonnywyatt — APPROVED** [review]
